File: wbs/wbs/report/wbs_stock_ledger_report/wbs_stock_ledger_report.py
# Copyright (c) 2013, yashwanth and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from frappe import _
from frappe.utils import cint, flt
from erpnext.stock.utils import update_included_uom_in_report
from erpnext.stock.doctype.serial_no.serial_no import get_serial_nos

def execute(filters=None):
	data = []
	columns = get_columns()
	items = get_items(filters)
	sl_entries = get_stock_ledger_entries(filters, items)

	return columns, data


def get_columns():
	columns = [
		{"label": _("Date"), "fieldname": "date", "fieldtype": "Datetime", "width": 95},
		{"label": _("Item"), "fieldname": "item_code", "fieldtype": "Link", "options": "Item", "width": 130},
		{"label": _("Item Name"), "fieldname": "item_name", "width": 100},
		{"label": _("Item Group"), "fieldname": "item_group", "fieldtype": "Link", "options": "Item Group", "width": 100},
		{"label": _("Brand"), "fieldname": "brand", "fieldtype": "Link", "options": "Brand", "width": 100},
		{"label": _("Description"), "fieldname": "description", "width": 200},
		{"label": _("Warehouse"), "fieldname": "warehouse", "fieldtype": "Link", "options": "Warehouse", "width": 100},
		{"label": _("WBS Storage Location"), "fieldname": "wbs_storage_location", "fieldtype": "Link", "options": "WBS Storage Location", "width": 100},
		{"label": _("Stock UOM"), "fieldname": "stock_uom", "fieldtype": "Link", "options": "UOM", "width": 100},
		{"label": _("Qty"), "fieldname": "actual_qty", "fieldtype": "Float", "width": 50, "convertible": "qty"},
		{"label": _("Balance Qty"), "fieldname": "qty_after_transaction", "fieldtype": "Float", "width": 100, "convertible": "qty"},
		{"label": _("Incoming Rate"), "fieldname": "incoming_rate", "fieldtype": "Currency", "width": 110,
			"options": "Company:company:default_currency", "convertible": "rate"},
		{"label": _("Valuation Rate"), "fieldname": "valuation_rate", "fieldtype": "Currency", "width": 110,
			"options": "Company:company:default_currency", "convertible": "rate"},
		{"label": _("Balance Value"), "fieldname": "stock_value", "fieldtype": "Currency", "width": 110,
			"options": "Company:company:default_currency"},
		{"label": _("Voucher Type"), "fieldname": "voucher_type", "width": 110},
		{"label": _("Voucher #"), "fieldname": "voucher_no", "fieldtype": "Dynamic Link", "options": "voucher_type", "width": 100},
		{"label": _("Batch"), "fieldname": "batch_no", "fieldtype": "Link", "options": "Batch", "width": 100},
		{"label": _("Serial No"), "fieldname": "serial_no", "width": 100},
		{"label": _("Balance Serial No"), "fieldname": "balance_serial_no", "width": 100},
		{"label": _("Project"), "fieldname": "project", "fieldtype": "Link", "options": "Project", "width": 100},
		{"label": _("Company"), "fieldname": "company", "fieldtype": "Link", "options": "Company", "width": 110}
	]

	return columns


def get_items(filters):
	conditions = []
	if filters.get("item_code"):
		conditions.append("item.name=%(item_code)s")
	else:
		if filters.get("brand"):
			conditions.append("item.brand=%(brand)s")
		if filters.get("item_group"):
			conditions.append(get_item_group_condition(filters.get("item_group")))

	items = []
	if conditions:
		items = frappe.db.sql_list("""select name from `tabItem` item where {}"""
			.format(" and ".join(conditions)), filters)
	return items

def get_stock_ledger_entries(filters, items):
	item_conditions_sql = ''
	if items:
		item_conditions_sql = 'and sle.item_code in ({})'\
			.format(', '.join([frappe.db.escape(i) for i in items]))

	return frappe.db.sql("""select concat_ws(" ", posting_date, posting_time) as date,
			item_code, warehouse, actual_qty, qty_after_transaction, incoming_rate, valuation_rate,
			stock_value, voucher_type, voucher_no, batch_no, serial_no, company, project, stock_value_difference
		from `tabStock Ledger Entry` sle
		where company = %(company)s and
			posting_date between %(from_date)s and %(to_date)s
			{sle_conditions}
			{item_conditions_sql}
			order by posting_date asc, posting_time asc, creation asc"""\
		.format(
			sle_conditions=get_sle_conditions(filters),
			item_conditions_sql = item_conditions_sql
		), filters, as_dict=1)

File: wbs/wbs/report/wbs_stock_ledger_report/wbs_stock_ledger_report.js
// Copyright (c) 2016, yashwanth and contributors
// For license information, please see license.txt
/* eslint-disable */
var to_date;
var from_date;

frappe.query_reports["WBS Stock Ledger Report"] = {
	"filters": [
		{
			"fieldname":"company",
			"label": __("Company"),
			"fieldtype": "Link",
			"options": "Company",
			"default": frappe.defaults.get_user_default("Company"),
			"reqd": 1
		},
		{
			"fieldname":"wbs_settings",
			"label": __("WBS Settings"),
			"fieldtype": "Link",
			"options": "WBS Settings",
			"reqd": 1,
			"on_change": function() {
				let id = frappe.query_report.get_filter_value('wbs_settings')
				if (id) {
					let start_date = get_start_date(id)
					let end_date = get_end_date(id)
					let warehouse = get_warehouse(id)
					if (start_date) {
						from_date = start_date
						frappe.query_report.set_filter_value('from_date', start_date)
					}

					if (end_date) {
						to_date = end_date
						frappe.query_report.set_filter_value('to_date', end_date)
					}

					if (warehouse) {
						frappe.query_report.set_filter_value('warehouse', warehouse)
					}
				}
			}
		},
		{
			"fieldname":"from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"reqd": 1
		},
		{
			"fieldname":"to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"reqd": 1
		},
		{
			"fieldname":"warehouse",
			"label": __("Warehouse"),
			"fieldtype": "Link",
			"options": "Warehouse",
			"get_query": function() {
				const company = frappe.query_report.get_filter_value('company');
				return {
					filters: { 'company': company }
				}
			}
		},
		{
			"fieldname":"item_code",
			"label": __("Item"),
			"fieldtype": "Link",
			"options": "Item",
			"get_query": function() {
				return {
					query: "erpnext.controllers.queries.item_query"
				}
			}
		},
		{
			"fieldname":"item_group",
			"label": __("Item Group"),
			"fieldtype": "Link",
			"options": "Item Group"
		},
		{
			"fieldname":"batch_no",
			"label": __("Batch No"),
			"fieldtype": "Link",
			"options": "Batch"
		},
		{
			"fieldname":"brand",
			"label": __("Brand"),
			"fieldtype": "Link",
			"options": "Brand"
		},
		{
			"fieldname":"voucher_no",
			"label": __("Voucher #"),
			"fieldtype": "Data"
		},
		{
			"fieldname":"project",
			"label": __("Project"),
			"fieldtype": "Link",
			"options": "Project"
		},
		{
			"fieldname":"include_uom",
			"label": __("Include UOM"),
			"fieldtype": "Link",
			"options": "UOM"
		}
	]
};

function get_start_date(id) {
	let date;
	frappe.call({
		method: 'wbs.wbs.doctype.wbs_settings.wbs_settings.get_start_date',
		args: {
			'ID': id
		},
		async: false,
		callback: (r) => {
			if(r.message.from_date) {
				date = r.message.from_date
			} else if (r.message.EX){
				frappe.throw(__(r.message.EX))
			} else {
				date = false
			}
		}
	});
	return date
}

function get_end_date(id) {
	let date;
	frappe.call({
		method: 'wbs.wbs.doctype.wbs_settings.wbs_settings.get_end_date',
		args: {
			'ID': id
		},
		async: false,
		callback: (r) => {
			if(r.message.to_date) {
				date = r.message.to_date
			} else if (r.message.INFINITE) {
				date = ''
			} else if (r.message.EX){
				frappe.throw(__(r.message.EX))
			} else {
				date = false
			}
		}
	});
	return date
}

function get_warehouse(id){
	let warehouse;
	frappe.call({
		method: 'wbs.wbs.doctype.wbs_settings.wbs_settings.get_warehouse',
		args: {
			'ID': id
		},
		async: false,
		callback: (r) => {
			if (r.message.warehouse) {
				warehouse = r.message.warehouse
			} else if (r.message.EX) {
				frappe.throw(__(r.message.EX))
			} else {
				warehouse = false
			}
		}
	})
	return warehouse
}
